source/common/thread_local/thread_local_impl.cc

```diff
@@ -26,79 +26,73 @@ SlotPtr InstanceImpl::allocateSlot() {
   ASSERT(!shutdown_);
 
   if (free_slot_indexes_.empty()) {
-    SlotImplPtr slot(new SlotImpl(*this, slots_.size()));
-    auto wrapper = std::make_unique<Bookkeeper>(*this, std::move(slot));
-    slots_.push_back(wrapper->slot_.get());
-    return wrapper;
+    SlotPtr slot(new SlotImpl(*this, slots_.size()));
+    slots_.push_back(slot.get());
+    return slot;
   }
   const uint32_t idx = free_slot_indexes_.front();
   free_slot_indexes_.pop_front();
   ASSERT(idx < slots_.size());
-  SlotImplPtr slot(new SlotImpl(*this, idx));
+  SlotPtr slot(new SlotImpl(*this, idx));
   slots_[idx] = slot.get();
-  return std::make_unique<Bookkeeper>(*this, std::move(slot));
+  return slot;
 }
 
-bool InstanceImpl::SlotImpl::currentThreadRegistered() {
-  return thread_local_data_.data_.size() > index_;
-}
+InstanceImpl::SlotImpl::SlotImpl(InstanceImpl& parent, uint32_t index)
+    : parent_(parent), index_(index), still_alive_guard_(std::make_shared<bool>(true)) {}
 
-void InstanceImpl::SlotImpl::runOnAllThreads(const UpdateCb& cb) {
-  parent_.runOnAllThreads([this, cb]() { setThreadLocal(index_, cb(get())); });
+Event::PostCb InstanceImpl::SlotImpl::wrapCallback(Event::PostCb cb) {
+  // See the header file comments for still_alive_guard_ for the purpose of this capture.
+  return [still_alive_guard = std::weak_ptr<bool>(still_alive_guard_), cb] {
+    if (!still_alive_guard.expired()) {
```

Comment:
is my understanding correct now that we dont defer deletion of slot in main thread, we lost the safe guard on cb() accessing the soon deleted slot after passing the still_alive_guard?

Reply:
> is my understanding correct now that we dont defer deletion of slot in main thread, we lost the safe guard on cb() accessing the soon deleted slot after passing the still_alive_guard?

Sorry I can't full parse this. Rephrase? No deferred deletion is required anymore because we don't capture anything that is owned by the slot. Thus, the still alive guard will be immediately freed and any future checks on workers will fail. This doesn't prevent bad captures (see comments) and races, but this is no different from before. (It's better than before because callbacks will not run in man cases that have previously caused crashes.)

Reply:
sorry, I am thinking of this scenario: 
1. if (!still_alive_guard.expired()) passes.
2. cb() run in workers
3. slot destructs in main thread, (now still_alive_guard.expired). and cb access slot(), this leads to undefined behavior?

Reply:
Why would the callback access the slot? I don't think this makes sense.

Reply:
yeah, we probably need some refactor to get them out of the cb. there are a few of them, for example: https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/cluster_manager_impl.cc#L654

Reply:
OK thanks I will fix.

Reply:
In order to fix this, I'm just removing the PostCb APIs, and switching the the UpdateCb style APIs which pass the current value, and it's actually forcing fixing a bunch of broken captures, so great suggestion!

/wait

Reply:
Thanks, I think the cleanup makes lots of sense, and much cleaner now!